lib/grpcServer/handlers/getTransactionsByFilterHandlerFactory.js

```diff
@@ -1,12 +1,104 @@
-const { RawTransaction } = require('@dashevo/dapi-grpc');
+const BloomFilter = require('bloom-filter');
+const { MerkleBlock } = require('@dashevo/dashcore-lib');
+const { TransactionFilterResponse } = require('@dashevo/dapi-grpc');
 
-function getTransactionsByFilterHandlerFactory() {
+const TransactionHashesCache = require('../../transactionsFilter/TransactionHashesCache');
+const BloomFilterEmitter = require('../../bloomFilter/emitter/BloomFilterEmitter');
+
+/**
+ *
+ * @param {BloomFilterEmitterCollection} bloomFilterEmitterCollection
+ * @param {testFunction} testTransactionAgainstFilter
+ * @return {getTransactionsByFilterHandler}
+ */
+function getTransactionsByFilterHandlerFactory(
+  bloomFilterEmitterCollection,
+  testTransactionAgainstFilter,
+) {
+  /**
+   * @typedef getTransactionsByFilterHandler
+   * @param {Object} call
+   * @param {function(Error, Object)} callback
+   */
   function getTransactionsByFilterHandler(call, callback) {
-    const transaction = new RawTransaction();
+    // Create a new bloom filter emitter when client connects
+
+    let filter;
+    try {
+      filter = new BloomFilter(call.request);
+    } catch (e) {
+      const grpcError = new Error(`Invalid bloom filter: ${e.message}`);
+
+      callback(grpcError, null);
+
+      return;
+    }
+
+    const filterEmitter = new BloomFilterEmitter(filter, testTransactionAgainstFilter);
+
+    const matchedTransactionHashes = new TransactionHashesCache();
+
+    // Send a raw transaction or a locked transaction hash when they match the bloom filter
+    filterEmitter.on('match', (transaction) => {
+      const response = new TransactionFilterResponse();
+
+      // Store the matched transaction hash
+      // in order to build a merkle block with sent transactions
+      matchedTransactionHashes.add(transaction.hash);
+
+      // Send transaction
+      response.setRawTransaction(transaction.toBuffer());
+
+      call.write(response.toObject());
+    });
+
+    // Send merkle block with previously sent transactions when they mined
+    filterEmitter.on('block', (block) => {
+      const blockTransactionHashes = block.transactions.map(t => t.hash);
+
+      // Remove transactions from cache if they have enough confirmations
+      matchedTransactionHashes.updateByBlockTransactionHashes(blockTransactionHashes);
+
+      // Mark matched transactions in the block
+      let hasMatchedTransactions = false;
+      const matchedTransactionFlags = blockTransactionHashes.map((hash) => {
+        const isMatched = matchedTransactionHashes.getConfirmationsCount(hash) === 1;
+
+        if (isMatched) {
```

Comment:
`const isMatched = matchedTransactionHashes.getConfirmationsCount(hash) === 1`

I think you are marking transactions that have their hashes stored in the previously matched unconfirmed transaction hashes cache `matchedTransactionHashes `. It would be great for performance if it were that simple, but unfortunately we cannot assume that all matching transactions come to the node first as unconfirmed single transactions. It can happen that the node gets to know the matching transaction hash first from the newly mined block. E.g if the miner includes an unknown transaction in the block and the network learns about it first through block propagation. Unfortunately this means we will have to test filters against every transaction in the block.

Reply:
Interesting.. I totally missed that. Thank you. I see a few options for how we can deal with that:
1. Store unmatched transaction hashes in cache as well. Then we can test only transaction which didn't appear before. It will take more memory.
2. Listen block hash then get block with transactions via RPC then test all of them. It will take additional bandwidth and CPU.

I like the first one. As I saw we have 100 transactions per block max and we are storing only hashes. We could use Buffer instead of hex just to reduce memory size.

Reply:
Awesome idea, I like number 1. as well!

Reply:
Probably, `txraw` event will be emitted before `rawblock` https://dashpay.slack.com/archives/G7W56UMCG/p1559563325018600

Reply:
Fortunately, we don't have this problem. All `tx` events fire before `block` event.

Reply:
> Fortunately, we don't have this problem. All `tx` events fire before `block` event.

That's great. Thanks for clearing it up!